drake/systems/DrakeSystem.h

```diff
@@ -0,0 +1,124 @@
+#ifndef __DrakeSystem_H__
+#define __DrakeSystem_H__
+
+#include <string>
+#include <memory>
+#include <stdexcept>
+#include <Eigen/Dense>
+#include "CoordinateFrame.h"
+
+#undef DLLEXPORT
+#if defined(WIN32) || defined(WIN64)
+#if defined(drakeSystem_EXPORTS)
+#define DLLEXPORT __declspec( dllexport )
+#else
+#define DLLEXPORT __declspec( dllimport )
+#endif
+#else
+#define DLLEXPORT
+#endif
+
+
+/// A dynamical system authored in Drake registers it's dynamics as well as information
+/// about it's coordinate frames.
+
+class DLLEXPORT DrakeSystem : public std::enable_shared_from_this<DrakeSystem> {
+public:
+
+  typedef Eigen::Matrix<double,Eigen::Dynamic,1> VectorXs;
+
+  DrakeSystem(const std::string& name,
+              const std::shared_ptr<CoordinateFrame>& continuous_state_frame,  // note: the frames are not const
+              const std::shared_ptr<CoordinateFrame>& discrete_state_frame,
+              const std::shared_ptr<CoordinateFrame>& input_frame,
+              const std::shared_ptr<CoordinateFrame>& output_frame);
+
+  DrakeSystem(const std::string& name,
+              unsigned int num_continuous_states = 0,
+              unsigned int num_discrete_states = 0,
+              unsigned int num_inputs = 0,
+              unsigned int num_outputs = 0);
+
+  virtual ~DrakeSystem(void) {};
+
+  const CoordinateFrame& getInputFrame() { return *input_frame.get(); }
+  const CoordinateFrame& getStateFrame() { return *state_frame.get(); }
+  const CoordinateFrame& getOutputFrame() { return *output_frame.get(); }
+
+  // todo: templates for these
+
+  virtual VectorXs dynamics(double t, const VectorXs& x, const VectorXs& u) {
+    throw std::runtime_error("(Drake:DrakeSystem:dynamics) systems with continuous dynamics must overload the dynamics method");
+  }
+
+  virtual VectorXs update(double t, const VectorXs& x, const VectorXs& u) {
+    throw std::runtime_error("(Drake:DrakeSystem:update) systems with discrete dynamics must overload the update method");
+  }
+
+  virtual VectorXs output(double t, const VectorXs& x, const VectorXs& u) {
+    if (output_frame->getDim()>0)
+      throw std::runtime_error("(Drake:DrakeSystem:dynamics) systems with outputs must overload the output method");
+    else
+      return VectorXs::Zero(0);
+  }
+
+  virtual VectorXs getRandomState(void);
+  virtual VectorXs getInitialState(void);
+
+  // simulation options
+  typedef struct _SimulationOptions {
+    double realtime_factor;  // 1 means try to run at realtime speed, < 0 is run as fast as possible
+    double initial_step_size;
+
+    _SimulationOptions(void) :
+            realtime_factor(-1.0),
+            initial_step_size(0.01)
+    {};
+  } SimulationOptions;
+  SimulationOptions default_simulation_options;
+
+  virtual void simulate(double t0, double tf, const VectorXs& x0, const SimulationOptions& options);
+  virtual void simulate(double t0, double tf, const VectorXs& x0) {
+    simulate(t0,tf,x0,default_simulation_options);
+  }
```

Comment:
i’m fine with that in the long run, but think we can put this in as is to get going.

> On Aug 22, 2015, at 9:21 PM, Twan Koolen <notifications@github.com> wrote:
> 
> In drake/systems/DrakeSystem.h <https://github.com/RobotLocomotion/drake/pull/1274#discussion_r37701586>:
> 
> > +  // simulation options
> > +  typedef struct _SimulationOptions {
> > +    double realtime_factor;  // 1 means try to run at realtime speed, < 0 is run as fast as possible
> > +    double initial_step_size;
> > +
> > +    _SimulationOptions(void) :
> > +            realtime_factor(-1.0),
> > +            initial_step_size(0.01)
> > +    {};
> > +  } SimulationOptions;
> > +  SimulationOptions default_simulation_options;
> > +
> > +  virtual void simulate(double t0, double tf, const VectorXs& x0, const SimulationOptions& options);
> > +  virtual void simulate(double t0, double tf, const VectorXs& x0) {
> > +    simulate(t0,tf,x0,default_simulation_options);
> > +  }
> Maybe have a separate Simulator class with a constructor that takes a DrakeSystem, SimulationOptions, and an Integrator or something, which has a simulate method. I feel like that's a nice piece of functionality we can break off and put in a different file, instead of having it clutter DrakeSystem.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/RobotLocomotion/drake/pull/1274/files#r37701586>.
>